utils/training.py

```diff
@@ -133,6 +136,125 @@ def train_shakespeare_transformer(model, train_loader, eval_loader, optimizer, n
     plt.show()
 
 
+def train_nanogpt(model, device, train_data_loader, val_data_loader, max_iters, batch_size):
+    device_type = 'cuda' if 'cuda' in str(device) else 'cpu'
+    weight_decay=1e-1
+    dtype = 'bfloat16' if torch.cuda.is_available() and torch.cuda.is_bf16_supported() else 'float16' # 'float32', 'bfloat16', or 'float16', the latter will auto implement a GradScaler
+    learning_rate = 6e-4 # max learning rate
+    beta1=0.9
+    beta2=0.95
+    grad_clip=1.0
+    eval_iters = 200
+    ptdtype = {'float32': torch.float32, 'bfloat16': torch.bfloat16, 'float16': torch.float16}[dtype]
+    ctx = nullcontext() if device_type == 'cpu' else torch.amp.autocast(device_type=device_type, dtype=ptdtype)
+    warmup_iters = max(1, int(max_iters / 30.0)) # how many steps to warm up for
+    lr_decay_iters = max_iters # should be ~= max_iters per Chinchilla
+    min_lr = 6e-5 # minimum learning rate, should be ~= learning_rate/10 per Chinchilla
+    decay_lr = True # whether to decay the learning rate
+    master_process = True
+    eval_only = False 
+    gradient_accumulation_steps = 5 * 8 # used to simulate larger batch sizes
+    log_interval = 1
+    iter_num = 0
+
+    scaler = torch.cuda.amp.GradScaler(enabled=(dtype == 'float16'))
+    optimizer = model.configure_optimizers(weight_decay, learning_rate, (beta1, beta2), device_type)
+
+    def get_batch(split):
+        loader = train_data_loader if split == 'train' else val_data_loader
+
+        # randomly select a batch of sequences
+        for batch_idx, (input_chunk, target_chunk) in enumerate(loader):
+            x, y = input_chunk.to(device_type), target_chunk.to(device_type)
+            return x, y
+
+    # helps estimate an arbitrarily accurate loss over either split using many batches
+    @torch.no_grad()
+    def estimate_loss():
+        out = {}
+        model.eval()
+        for split in ['train', 'val']:
+            losses = torch.zeros(eval_iters)
+            for k in range(eval_iters):
+                X, Y = get_batch(split)
```

Comment:
I think we need to add the logic here to split the chunk into a piece of context_size, right?